### PR TITLE
Introduce custom caching for authorization and authentication info

### PR DESCRIPTION
**Pull request by fickludd**
- Turn off Shiro session storage
- Replace `EhCache` with custom cache management
- Expose auth cache TTL and maxCapacity as configurable settings
- Turn off authorization info caching for InternalFlatFileRealm
- Turn on authentication and authorization info caching for LdapRealm

### CODE REVIEW

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthCache.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    @Override
+    public V get( K key ) throws CacheException
+    {
+        return get( key, ignored -> {} );
+    }
+
+    V get( K key, final Consumer<Boolean> onComplete ) throws CacheException
+    {
+        if ( needsValidation() )
+        {
```
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthCache.java]: This could be a bit of a resource hog on smaller TTL settings. Maybe you can relax the 5% heuristic since you are checking the TTL of the entry you are getting anyways?

Also you do not want to queue more that one validation job at any time, since in bad cases you could end up with a flood of meaningless validation jobs.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthCache.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import org.neo4j.helpers.collection.Pair;
+
+import static java.util.stream.Collectors.toList;
+
+public class AuthCache<K, V> implements Cache<K,V>
+{
+    private final ConcurrentLinkedHashMap<K,Pair<Long,V>> cacheMap;
+    private final Clock clock;
+    private final long TTL;
+    private long lastValidation;
```
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthCache.java]: This is accessed from both threads so should be an AtomicLong

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/SecuritySettings.java [hunk 3]
```diff
@@ -130,4 +132,12 @@
                   "E.g. group1=role1;group2=role2;group3=role3,role4,role5" )
     public static Setting<String> ldap_authorization_group_to_role_mapping =
             setting( "dbms.security.realms.ldap.authorization.group_to_role_mapping", STRING, NO_DEFAULT );
+
+    @Description( "The time to live (TTL) for cached authentication and authorization info." )
+    public static Setting<Long> auth_caching_ttl =
```
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/SecuritySettings.java]: `auth_cache_ttl` is more consistent with the next setting (and better to change this than have auth_caching_max_capacity which sounds weird)

File: enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/InternalFlatFileRealmTest.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    private class TestRealm extends InternalFlatFileRealm
+    {
+        private boolean authenticated = false;
+        private boolean authorized = false;
+
+        boolean wasAuthenticated()
+        {
+            boolean t = authenticated;
+            authenticated = false;
```
Comment (henriknyman) [inline enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/InternalFlatFileRealmTest.java]: I found these tests slightly hard to follow because of these side-effects (here and wasAuthorized()). Maybe we could make it more explicit?

File: enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthCacheTest.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+public class AuthCacheTest
+{
+    private AuthCache<Integer,String> cache;
+    private FakeClock fakeClock;
+
+    @Before
+    public void setUp()
+    {
+        fakeClock = Clocks.fakeClock();
+        cache = new AuthCache<>( fakeClock, 100, 5 );
```
Comment (henriknyman) [inline enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthCacheTest.java]: There are no tests for TTL = 0

**Comment by lutovich** [comment]
@fickludd we have ehcache in the classpath (transitive dependency of shiro-ehcache). can we use some existing cache implementation from there instead of building our own?

**Comment by lutovich** [comment]
@fickludd hm, i did not notice that shiro-ehcache was removed in one of commits. why can't we use it or some other caching lib instead of writing our own?

_Update:_ please also consider `LruCache`, `ClockCache` and `TimedRepository` in our codebase.

**Comment by Mats-SX** [comment]
@lutovich To inject the cache we need to implement a `Cache` interface given by Shiro. Because of that we must have a (thin) wrapper around some other implementation, and we chose a `ConcurrentLinkedHashmap`, which is already a dependency (used for plan caching in Cypher). We could choose to make a thin wrapper around `TimedRepository` instead, as I can see it allows you to control time by choosing the `Clock` implementation from outside. Do you think that would be better than our current wrapper?

We specifically do not want `EhCache`, because it is overly cumbersome for our use-case. Also there was a problem discovered by the big-graph team caused by it.

File: enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/InternalFlatFileRealmTest.java (hunk 4 above)
Comment (Mats-SX) [reply enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/InternalFlatFileRealmTest.java]: Do you mean the reset side-effect?

**Comment by lutovich** [comment]
@Mats-SX current cache implementation is slightly more than a wrapper - it has thread and ttl tracking, for example. i think using `TimedRepository` might result in less code being written, which is nice.

**Comment by fickludd** [comment]
@lutovich Thanks for pointing out our already numerous cache implementations. We re-evaluated the choice of an auth cache in the security team today, and decided to go with the caffeine cache instead, which has all the functionality we need. None of the 3 neo caches really cover our use case - we need ttl from write. We still need a wrapper to join it with Shiro though. The caffeine cache is not in our code-base yet, but it replaces the ConcurrentLinkedHashMap that is used for Query planning.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    {
+        return cacheMap.getIfPresent( key );
+    }
+
+    @Override
+    public V put( K key, V value ) throws CacheException
+    {
+        V oldValue = cacheMap.getIfPresent( key );
+        cacheMap.put( key, value );
+        return oldValue;
```
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: The Caffeine docs seem to prescribe using the `V get(@Nonnull K key, @Nonnull Function<? super K, ? extends V> mappingFunction);` method to perform this kind of update atomically

Comment (fickludd) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: We considered this, but it's really not what we need. That get does `if cached, return; otherwise create, cache and return`while we need `if cached, return old and cache new; otherwise cache new and return null`.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    @Override
+    public V remove( K key ) throws CacheException
+    {
+        V oldValue = cacheMap.getIfPresent( key );
+        cacheMap.invalidate( key );
+        return oldValue;
```
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: It would be nice if this had an atomic version as well, but it also looks like Shiro doesn't care about these return values so it may not be that important.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java (hunk 6 above)
Comment (ben-manes) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: This can be done atomically as a single operation if you use the `asMap()` view.

``` java
return cache.asMap().put(key, value);
```

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java (hunk 7 above)
Comment (ben-manes) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: Please consider using the `asMap()` view.

``` java
return cache.asMap().remove(key);
```

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    @Override
+    public void clear() throws CacheException
+    {
+        cacheMap.invalidateAll();
+    }
+
+    @Override
+    public int size()
+    {
+        return Math.toIntExact( cacheMap.estimatedSize() );
```
Comment (ben-manes) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: alternatively `cache.asMap().size()` will saturate at `Integer.MAX_VALUE`

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            this.maxCapacity = maxCapacity;
+            cacheMap = new HashMap<>();
+        }
+
+        @Override
+        public <K, V> Cache<K,V> getCache( String s ) throws CacheException
+        {
+            if ( !cacheMap.containsKey( s ) )
+            {
+                cacheMap.put( s, ttl <= 0 ? new NullCache() : new ShiroCaffeineCache<K,V>( ticker, ttl, maxCapacity ) );
```
Comment (ben-manes) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: perhaps use an atomic version, e.g. 

``` java
return cacheMap.computeIfAbsent(s, key -> {
  return ...
});
```

File: enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCacheTest.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    public void shouldRemoveIfExceededCapacity()
+    {
+        cache.put( 1, "one" );
+        cache.put( 2, "two" );
+        cache.put( 3, "three" );
+        cache.put( 4, "four" );
+        cache.put( 5, "five" );
+        cache.put( 6, "six" );
+
+        assertThat( cache.keys(), containsInAnyOrder( 2, 3, 4, 5, 6 ) );
```
Comment (ben-manes) [inline enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCacheTest.java]: Because the algorithm isn't LRU and has some randomness to protect against HashDoS attacks, this could be flaky. The random hash seed could cause a collision in the frequency sketch resulting in `1` being retained. You might prefer to check the size instead of the contents.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java (hunk 6 above)
Comment (henriknyman) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: Right, I didn't read carefully. Thanks for popping in @ben-manes!

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java (hunk 8 above)
Comment (fickludd) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: @ben-manes thanks for the tips! It seems though that `cache.asMap().size()` and `cache.asMap().keySet().size()` do not necessarily yield the same value... is that expected? Noticed this in the `shouldRemoveIfExceededCapacity()` test.

File: enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCacheTest.java (hunk 10 above)
Comment (fickludd) [reply enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCacheTest.java]: Aha, this was a remainder test from the previous backing cache. We should have read the caffeine documentation better for sure!

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java (hunk 8 above)
Comment (ben-manes) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: @fickludd, you're observing the [asynchronous maintenance](https://github.com/ben-manes/caffeine/wiki/Testing). In CLHM and Guava the work is amortized on the caller's thread, as the penalty is low and we don't want our own threads. In Java 8 we delegate it to `ForkJoinPool.commonPool()` by default, which is shared and lets us minimize user-facing latencies. That makes testing a little more complicated than it was previously.

You can either specify a same-thread executor in tests (`Caffeine.newBuilder().executor(Runnable::run)`) or call `Cache.cleanUp()`. I prefer the executor. With that change your test passes :)

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java (hunk 9 above)
Comment (ben-manes) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: Oh this wasn't a `ConcurrentMap`, so there was no difference. Is this ever called concurrently or only during initialization?

Comment (fickludd) [reply enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCache.java]: It's called twice during start-up, with different arguments, so I think it should be safe from nasty surprises. But your comment made change the variable names, because they were slightly confusing.
